### PR TITLE
8: Adds support for system-versioned temporal tables

### PR DESCRIPTION
**Pull request by jlhogan**
- Creates DbSetExtensions for querying temporal tables
- Creates MigrationBuilderExtensions for adding temporal table defn to migrations
- Updates Child table to be temporal
- Updates Child Repository to enable querying for a single child by Id and asOf timestamp (GetChildByIdAsOfAsync)
- Creates GraphQL query to expose GetChildByIdAsOfAsync
- Adds a docker-compose.test.yaml file for running tests

Wanted to get this up for discussion; will continue to implement for other immutable data types on a new branch. 

Question for reviewers:
Bootstrapping for integration tests is handled by `TestClientProvider` which ultimately only returns a client to the test. Integration tests can seed DB backing the client by passing in a seed function to the TestClientProvider constructor. I'm considering if this should be refactored to make the context directly available to the test. This would avoid the pre-declaring of variables that will be set with values from the seed data that is currently happening in integration tests, but feels like it increases the potential for tests to get confusing and mess with stuff. Interested in your thoughts

Riders:
- some docker-compose file clean up

### CODE REVIEW

**Comment by cailyncodes** [comment]
Thanks for doing all this!!! I'm excited to try running it once I get back to my laptop. It seems very slick and so happy we didn't have to roll our own solution!

I really like the structure of the integration tests. Yes it's a little clunky with the pre variable declarations, but that's a tradeoff I'm willing to make given the flexibility it provides to customizing the seed data. It also still is very readable which I care the most about.

It looks like the build is failing though, do you have an idea why? (My phone isn't letting me open the Azure build pipeline info)

**Comment by jlhogan** [comment]
Build is failing b/c i forgot to add my update to the azure-pipeline defn! Added my commit to this branch, it's running now; will probably need a little futzing but should have it working momentarily :) 

glad you're finding it readable!! that is always my goal

File: src/Hedwig/Migrations/20190909160144_InitialCreate.cs [hunk 1]
```diff
@@ -8,6 +8,7 @@ public partial class InitialCreate : Migration
     {
         protected override void Up(MigrationBuilder migrationBuilder)
         {
+            migrationBuilder.CreateHistorySchema();
```
[the remark this answers is not in the document]
Comment (jlhogan) [reply src/Hedwig/Migrations/20190909160144_InitialCreate.cs]: yup! I have been using: 
`docker-compose stop db; docker-compose rm -f dc; docker-compose up -d db` to get a fully wiped out and re-created db (when just stopping the container and restarting, data is persisted to the restarted db)

File: src/Hedwig/Schema/Queries/ChildQuery.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+namespace Hedwig.Schema.Queries
+{
+	public class ChildQuery : ObjectGraphType<object>, IAppSubQuery
+	{
+		public ChildQuery(IChildRepository repository)
+		{
+			Field<ChildType>(
+				"child",
+				arguments: new QueryArguments(
+					new QueryArgument<IdGraphType> { Name = "id" },
```
[the remark this answers is not in the document]
Comment (jlhogan) [reply src/Hedwig/Schema/Queries/ChildQuery.cs]: Good question! Looking.

File: test/HedwigTests/Integration/GraphQLTests.cs [hunk 3]
```diff
@@ -22,32 +25,88 @@ public async Task GraphQL_Exists()
...
+				childId = child.Id;
+				asOf = Utilities.GetAsOfWithSleep();
+				child.FirstName = updatedName;
+				context.SaveChanges();
+			}
+
+			using (var client = new TestClientProvider(seedData).Client) {
+				HttpResponseMessage responseCurrent = await client.GetGraphQLAsync(
+					$@"{{
+						child(id: ""{childId.Value}"") {{
```
[the remark this answers is not in the document]
Comment (jlhogan) [reply test/HedwigTests/Integration/GraphQLTests.cs]: To use normal double quote inside a verbatim string ( prepended modifier @  allows for writing multi-line strings), you have to use double double quotes https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/verbatim

File: src/Hedwig/Migrations/20190909160144_InitialCreate.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jlhogan) [reply src/Hedwig/Migrations/20190909160144_InitialCreate.cs]: yes. I have been using:
`docker-compose stop db; docker-compose rm -f dc; docker-compose up -d db` to completely wipe out and reinit the db. Then in the backend container execute `dotnet ef database update` again to re-apply migrations.

simply stopping & restarting the container results in data being persisted, so the migration table is populated, so the migrations will not be re-run.

File: src/Hedwig/Schema/Queries/ChildQuery.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jlhogan) [reply src/Hedwig/Schema/Queries/ChildQuery.cs]: No sure! will look

File: src/Hedwig/Migrations/20190909160144_InitialCreate.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cailyncodes) [reply src/Hedwig/Migrations/20190909160144_InitialCreate.cs]: Thank you for these steps!

File: test/HedwigTests/Integration/GraphQLTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jlhogan) [reply test/HedwigTests/Integration/GraphQLTests.cs]: No, seems int  parameters do not need to be quoted for GraphQL, and all other queries are using int Ids. Here, the Guid and timestamp are strings.

**Comment by jlhogan** [comment]
closes #8

**Comment by jlhogan** [comment]
closing in favor of: https://github.com/ctoec/ecis-experimental/pull/78

**Pull request closed without merging**